receiver/googlecloudpubsubreceiver/internal/log_entry.go

```diff
@@ -0,0 +1,574 @@
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package internal // import "github.com/open-telemetry/opentelemetry-collector-contrib/receiver/googlecloudpubsubreceiver/internal"
+
+import (
+	"bytes"
+	"context"
+	"encoding/hex"
+	stdjson "encoding/json"
+	jsoniter "github.com/json-iterator/go"
+
+	"errors"
+	"fmt"
+	"strconv"
+	"strings"
+	"sync"
+	"time"
+
+	"github.com/iancoleman/strcase"
+	"go.uber.org/zap"
+
+	"cloud.google.com/go/logging/apiv2/loggingpb"
+	"go.opentelemetry.io/collector/pdata/pcommon"
+	"go.opentelemetry.io/collector/pdata/plog"
+	"google.golang.org/genproto/googleapis/api/monitoredres"
+	"google.golang.org/protobuf/encoding/protojson"
+	"google.golang.org/protobuf/reflect/protoreflect"
+	"google.golang.org/protobuf/reflect/protoregistry"
+	"google.golang.org/protobuf/types/known/anypb"
+)
+
+
+var json = jsoniter.ConfigCompatibleWithStandardLibrary
+
+var invalidTraceID = [16]byte{}
+var invalidSpanID = [8]byte{}
+
+func cloudLoggingTraceToTraceIDBytes(trace string) [16]byte {
+	// Format: projects/my-gcp-project/traces/4ebc71f1def9274798cac4e8960d0095
+	lastSlashIdx := strings.LastIndex(trace, "/")
+	if lastSlashIdx == -1 {
+		return invalidTraceID
+	}
+	traceIDStr := trace[lastSlashIdx+1:]
+
+	return traceIDStrTotraceIDBytes(traceIDStr)
+}
+
+func traceIDStrTotraceIDBytes(traceIDStr string) [16]byte {
+	traceIDSlice := [16]byte{}
+	decoded, err := hex.Decode(traceIDSlice[:], []byte(traceIDStr))
+	if err != nil || decoded != 16 {
+		return invalidTraceID
+	}
+
+	return traceIDSlice
+}
+
+func spanIDStrToSpanIDBytes(spanIDStr string) [8]byte {
+	spanIDSlice := [8]byte{}
+	decoded, err := hex.Decode(spanIDSlice[:], []byte(spanIDStr))
+	if err != nil || decoded != 8 {
+		return invalidSpanID
+	}
+
+	return spanIDSlice
+}
+
+var desc protoreflect.MessageDescriptor
+var descOnce sync.Once
+
+func getLogEntryDescriptor() protoreflect.MessageDescriptor {
+	descOnce.Do(func() {
+		var logEntry loggingpb.LogEntry
+
+		desc = logEntry.ProtoReflect().Descriptor()
+	})
+
+	return desc
+}
+
+// TranslateLogEntry translates a JSON-encoded LogEntry message into a pair of
+// pcommon.Resource and plog.LogRecord, trying to keep as close as possible to
+// the semantic conventions.
+//
+// For maximum fidelity, the decoding is done according to the protobuf message
+// schema; this ensures that a numeric value in the input is correctly
+// translated to either an integer or a double in the output. It falls back to
+// plain JSON decoding if payload type is not available in the proto registry.
```

Comment:
Following the proto schema is responsible for a fair amount of the code here. It's arguably not worth the effort to avoid a few mistranslated numeric values. 🤷

Reply:
I just remembered that this isn't just about int vs float, it's also important for bytes, durations, and timestamps, all of which are both encoded as strings according to [the protobuf json mapping](https://protobuf.dev/programming-guides/proto3/#json). Actually checking that, it seems that all integers are to be encoded as strings, but can be decoded from either string or number.

Reply:
additionally there should be an optional metric for unresolvable type uris so that people can know to add them if needed.